### PR TITLE
feat: Update the service and state to consume the persistence layer method

### LINKED ISSUE #397: Update the service to save and retrieve connection record using connection recorder storage
- Create Own Unique ConnectionID (status - Pending) [Use - UUID] : 
- Create unique threadID (requestID) and store the mapping to connectionID

key : connectionID, val : {endpoint, recKey, inviationID}
key : protocolID=(myDID+threadID), val : connectionID

protocolID should be the hash.

### LINKED ISSUE #398: [Inviter] ExRequest Processing: Store data
- Retrieve recipient publicKey from envelope and get the invitation from Store.
- Create Own Unique ConnectionID (status - Pending) [Use - UUID] :
- Get the thread ID from exRequest and store the mapping to connectionID

key : connectionID, val : {theirDID, theirLabel,invitationID}
key : protocolID=(theirDID+threadID), val : connectionID

protocolID should be the hash.

### LINKED ISSUE #400: [Inviter] ExResponse Creation : Store data
- Get the connection object by connectionID and DID and store DID Doc.

key : connectionID, val : {...,myDID}

### LINKED ISSUE #401: [Invitee] ExResponse Processing: Store data
- from threadID fetch the connectionID and then connection.
- Update connection with theirDID

key : connectionID, val : {...,theirDID, theirLabel}
key : protocolID=(theirDID+threadID), val : connectionID

protocolID should be the hash.

### PR DESCRIPTION
**Pull request by talwinder50**
Bob -> Invitee:

Bob receives an invitation from Alice
Bob creates and stores the connection Record {ConnectionID->uuid ,ThreadID -> uuid, Recipient, ServiceEndpoint, state -> invited, Theirlabel: invitation )
Store the mapping of {MyNS+ThreadID to ConnectionID}
Sends an automatic Exchange Request to Alice (using the ThreadID created in the step 3. )
Update the connectionRecord { state -> requested, Mydid-> from request}
Alice -> Inviter

Alice create the connection Record using the ThreadID in the request plus TheirNS {ConnectionID->uuid ,ThreadID :thid, state -> requested, TheirDID->did from request, TheirLabel-> from request }
Mapping (TheirNS+Thid -> ConnectionID)
Alice prepares the exchange Response to Bob (using the ThreadID created in the step 3. )
Update the connectionRecord { state -> responded, MyDID -> from response )
Send exchange Response to Bob
Bob -> Invitee

Bob receives the Exchange Response
Fetch the connection Record using MyNS+threadID from response
Update connectionRecord {state -> responded, TheirDID->did from response (sig_data)}
Bob sends an Ack (using the ThreadID created in the step 3.)
Update the connectionRecord {state->completed}

See description of the issue for the detail
closes #397 closes #398 closes #400 closes #401

Signed-off-by: talwinder.kaur <talwinder.kaur@securekey.com>

### CODE REVIEW

File: pkg/client/didexchange/client_test.go [hunk 1]
```diff
@@ -136,20 +137,20 @@ func TestClient_CreateInvitationWithDID(t *testing.T) {
 }
 
 func TestClient_QueryConnectionByID(t *testing.T) {
+	connID := "id1"
```
Comment (sandrask) [inline pkg/client/didexchange/client_test.go]: if these are constants you can declare them as constants (for this test)

File: pkg/client/didexchange/client_test.go [hunk 2]
```diff
@@ -161,35 +162,47 @@ func TestClient_QueryConnectionByID(t *testing.T) {
...
-		connRec := &didexchange.ConnectionRecord{State: "complete", ConnectionID: "conn_id1", ThreadID: "thid1"}
-		connRecBytes, err := json.Marshal(connRec)
+		connRec := &didexchange.ConnectionRecord{ConnectionID: connID, ThreadID: threadID, State: "complete"}
+		connBytes, err := json.Marshal(connRec)
 		require.NoError(t, err)
-		require.NoError(t, s.Put("conn_id1", connRecBytes))
-		_, err = c.GetConnection("id1")
+		c := didexchange.NewConnectionRecorder(&s)
+		require.NoError(t, s.Put("conn_id1", connBytes))
+		_, err = c.GetConnectionRecord(threadID)
```
Comment (sandrask) [inline pkg/client/didexchange/client_test.go]: check for specific error here

File: pkg/didcomm/protocol/didexchange/service.go [hunk 3]
```diff
@@ -68,7 +69,10 @@ type provider interface {
 }
 
 type connectionStore interface {
-	GetConnectionRecord(connectionID string) (*ConnectionRecord, error)
+	saveConnectionRecord(record *ConnectionRecord) error
```
Comment (sandrask) [inline pkg/didcomm/protocol/didexchange/service.go]: Why do we have un-exported methods in an interface, let's discuss

File: pkg/didcomm/protocol/didexchange/service.go [hunk 4]
```diff
@@ -154,6 +163,10 @@ func (s *Service) HandleInbound(msg *service.DIDCommMsg) error {
 	if !current.CanTransitionTo(next) {
 		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
 	}
+	// trigger message events
```
Comment (sandrask) [inline pkg/didcomm/protocol/didexchange/service.go]: How is this relevant to your change?

File: pkg/didcomm/protocol/didexchange/service.go [hunk 5]
```diff
@@ -187,48 +208,45 @@ func (s *Service) HandleOutbound(msg *service.DIDCommMsg, destination *service.D
...
+		if err = s.update(msg.Msg.Header.Type, connectionRecord); err != nil {
 			return fmt.Errorf("failed to persist state %s %w", next.Name(), err)
 		}
+
 		logger.Infof("persisted the connection using %s and updated the state to %s",
-			msg.ThreadID, next.Name())
+			connectionRecord.ThreadID, next.Name())
 
-		if err := action(); err != nil {
+		if err := action(); err != nil { //nolint
```
Comment (sandrask) [inline pkg/didcomm/protocol/didexchange/service.go]: why is linter complaining?

File: pkg/didcomm/protocol/didexchange/service.go [hunk 6]
```diff
@@ -412,30 +434,24 @@ func threadID(didCommMsg *service.DIDCommMsg) (string, error) {
 }
 
 func (s *Service) currentState(thid string) (state, error) {
-	conn, err := s.connectionStore.GetConnectionRecord(thid)
+	connRec, err := s.connectionStore.GetConnectionRecordByNSThreadID(thid)
```
Comment (sandrask) [inline pkg/didcomm/protocol/didexchange/service.go]: rename function parameter to include namespace "ns" in it

File: pkg/didcomm/protocol/didexchange/service.go [hunk 7]
```diff
@@ -412,30 +434,24 @@ func threadID(didCommMsg *service.DIDCommMsg) (string, error) {
...
+func (s *Service) update(msgType string, connectionRecord *ConnectionRecord) error {
+	if msgType == ConnectionRequest && connectionRecord.State == stateNameRequested {
+		return s.connectionStore.saveNewConnectionRecord(connectionRecord)
 	}
-
-	// todo following function will be replaced by persistence connection store save connection record
-	err = s.store.Put("conn_"+thid, connRecBytes)
-	if err != nil {
-		return fmt.Errorf("failed to write to store: %s", err)
+	if msgType == ConnectionInvite && connectionRecord.State == stateNameInvited {
```
Comment (sandrask) [inline pkg/didcomm/protocol/didexchange/service.go]: you can combine the two with OR statement

File: pkg/didcomm/protocol/didexchange/service.go (hunk 4 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service.go]: This was picked wrongly in rebase. Removed

File: pkg/client/didexchange/client_test.go (hunk 1 above)
Comment (talwinder50) [reply pkg/client/didexchange/client_test.go]: done

File: pkg/didcomm/protocol/didexchange/service.go [hunk 8]
```diff
@@ -187,48 +208,45 @@ func (s *Service) HandleOutbound(msg *service.DIDCommMsg, destination *service.D
 }
 
 func (s *Service) handle(msg *message) error {
-	logger.Infof("entered into private handle didcomm message: %s ", msg)
+	logger.Infof("entered into private handle didcomm message: %s ", msg.ThreadID)
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/didexchange/service.go]: update the msg as this is changed to threadID (eg "message with threadID").

File: pkg/didcomm/protocol/didexchange/service.go [hunk 9]
```diff
@@ -187,48 +208,45 @@ func (s *Service) HandleOutbound(msg *service.DIDCommMsg, destination *service.D
...
 	}
 	logger.Infof("next valid state to transition -> %s ", next.Name())
-
 	for !isNoOp(next) {
-		// TODO change from thread id to connection id #397
-		// TODO pass invitation id #397
 		s.sendMsgEvents(&service.StateMsg{
 			Type: service.PreState, Msg: msg.Msg, StateID: next.Name(),
-			Properties: createEventProperties(msg.ThreadID, "")})
+			Properties: createEventProperties("", "")})
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/didexchange/service.go]: We need to reconsider where we are creating the connectionID. Without any correlation, the consumers will not be able to fetch connection details (if required). For now, can we add a TODO ?

Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service.go]: Added TODO

File: pkg/didcomm/protocol/didexchange/service.go (hunk 8 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service.go]: added

File: pkg/didcomm/protocol/didexchange/service.go (hunk 7 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service.go]: done

File: pkg/didcomm/protocol/didexchange/service.go (hunk 6 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service.go]: done

File: pkg/didcomm/protocol/didexchange/service.go (hunk 5 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service.go]: was not required, removed

File: pkg/didcomm/protocol/didexchange/service.go (hunk 3 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service.go]: removed the interface, i dont think it was required

File: pkg/client/didexchange/client_test.go (hunk 2 above)
Comment (talwinder50) [reply pkg/client/didexchange/client_test.go]: done
